homeassistant/components/notify/legacy.py

```diff
@@ -115,9 +118,12 @@ async def async_platform_discovered(
         """Handle for discovered platform."""
         await async_setup_platform(platform, discovery_info=info)
 
-    hass.data[NOTIFY_DISCOVERY] = discovery.async_listen_platform(
-        hass, DOMAIN, async_platform_discovered
-    )
+    # Setup dispatcher
+    if integrations and not hass.data[NOTIFY_DISCOVERY]["integrations"]:
```

Comment:
This won't work. We need to set up the dispatcher to be able to load platforms via discovery. That has nothing to do with already loaded platforms.

Reply:
I Agree: So what do you mean with:
> I suggest we update the discovery helper to return the unsubscribe callback. Then we can unsubscribe when resetting the platform.

Reply:
I think we can simply unsubscribe the dispatcher in `async_reset_platform`. It will be set up again in `async_setup`.

https://github.com/home-assistant/core/blob/49bc572d6d411770e0fb6cc17b798fd978406766/homeassistant/helpers/reload.py#L80-L81

There could potentially be a race since we await these calls, so we should make sure to unsubscribe at the end of `async_reset_platform` before exiting that function.

Reply:
That doesn't matter. We'll unsubscribe and subscribe the same dispatch target each time.

Reply:
Correct, but it could be that 2 resets arrive first before the first setup is completed

Reply:
I added a check to make sure there is no duplicate dispatcher

Reply:
I think you're right, but I don't think that's expected. Then we should consider guarding `component.async_reset_platform` and not call it before `component.async_setup` is done.

Reply:
Added a guard